src/cacao.ts

```diff
@@ -0,0 +1,104 @@
+import * as dagCbor from '@ipld/dag-cbor'
+import * as multiformats from 'multiformats'
+import * as Block from 'multiformats/block'
+import { sha256 as hasher } from 'multiformats/hashes/sha2'
+import { SiweMessage } from './siwe'
+
+export type Header = {
+  t: string
+}
+
+export type Payload = {
+  domain: string
+  iss: string
+  chainId: number
+  aud: string
+  version: string
+  nonce: string
+  iat: number
+  nbf?: number
+  exp?: number
+  statement?: string
+  requestId?: string
+  resources?: Array<string>
+}
+
+export type Signature = {
+  s: string
+}
+
+export type Cacao = {
+  h: Header
+  p: Payload
+  s: Signature
+}
+
+export interface CacaoBlock {
+  value: Cacao
+  cid: multiformats.CID
+  bytes: Uint8Array
+}
+
+class CACAO {
+  cacao: Cacao
+  constructor(siwe: string | SiweMessage) {
+    const siweMessage = typeof siwe === 'string' ? new SiweMessage(siwe) : siwe
+    if (!siweMessage.signature) {
+      throw new Error('SIWE Message has not been signed yet')
+    }
+    this.cacao = this.fromSiweMessage(siweMessage)
+  }
+
+  async toCacaoBlock(): Promise<CacaoBlock> {
+    const block = await Block.encode({
+      value: this.cacao,
+      codec: dagCbor,
+      hasher: hasher,
+    })
+    return block as CacaoBlock
```

Comment:
`Block.encode` can accept a generic argument, as well as `Block` itself. `type CacaoBlock = Block<Cacao>`, and you could use `Block.encode<Cacao>`. Type casting is fragile, as it forces a type while disregarding (with some restrictions, but it does not help much) an original type.

Reply:
interesting, thanks! will do :)

Reply:
Hm im unable to get this approach to work. 

I cant seem to define the type as Block<Cacao> because Block is a namespaced class, not a type. 
Also, Block.encode<Cacao> seems to give a weird error I don't understand. We can sync on this on Discord.